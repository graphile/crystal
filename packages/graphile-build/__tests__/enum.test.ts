const { GraphQLEnumType } = require("graphql");
const { buildSchema, defaultPlugins } = require("../");

function EnumPlugin(builder) {
  builder.hook("GraphQLObjectType:fields", (fields, build, context) => {
    const { extend, newWithHooks } = build;
    const {
      scope: { isRootQuery },
    } = context;
    if (!isRootQuery) {
      return fields;
    }
    const MyEnum = newWithHooks(
      GraphQLEnumType,
      {
        name: "MyEnum",
        values: {
          ONE: { value: 1 },
          TWO: { value: 2 },
          THREE: { value: 3 },
        },
      },
      {
        isMyEnum: true,
      },
    );
    return extend(fields, {
      enum: {
        type: MyEnum,
      },
    });
  });
  builder.hook("GraphQLEnumType:values", (values, build, context) => {
    const { extend } = build;
    const {
      scope: { isMyEnum },
    } = context;
    if (!isMyEnum) {
      return values;
    }
    return extend(values, {
      FOUR: { value: 4 },
    });
  });
  builder.hook("GraphQLEnumType:values:value", (value, build, context) => {
    const { extend } = build;
    const {
      scope: { isMyEnum },
    } = context;
    if (isMyEnum && value.value < 4) {
      return extend(value, {
        deprecationReason: "We no longer support numbers smaller than PI",
      });
    } else {
      return value;
    }
  });
}

test("generated schema", async () => {
  const schema = await buildSchema([...defaultPlugins, EnumPlugin]);
  expect(schema).toMatchSnapshot();
});
