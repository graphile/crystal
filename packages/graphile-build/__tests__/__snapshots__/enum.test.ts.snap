// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generated schema 1`] = `
enum MyEnum {
  ONE @deprecated(reason: "We no longer support numbers smaller than PI")
  TWO @deprecated(reason: "We no longer support numbers smaller than PI")
  THREE @deprecated(reason: "We no longer support numbers smaller than PI")
  FOUR
}

"""An object with a globally unique \`ID\`."""
interface Node {
  """
  A globally unique identifier. Can be used in various places throughout the system to identify this single value.
  """
  id: ID!
}

"""The root query type which gives access points into the data universe."""
type Query implements Node {
  """
  Exposes the root query type nested one level down. This is helpful for Relay 1
  which can only query top level fields if they are in a particular form.
  """
  query: Query!

  """
  The root query type must be a \`Node\` to work well with Relay 1 mutations. This just resolves to \`query\`.
  """
  id: ID!

  """Fetches an object given its globally unique \`ID\`."""
  node(
    """The globally unique \`ID\`."""
    id: ID!
  ): Node
  enum: MyEnum
}

`;
